Add error panel and tab to task details page

rsasch opened the pull request:
- [x] Display first few error messages (including task name and links to logs & execution dir) on the task details page
- [x] Add a tab to the "Tasks" tab group that displays all the errors' info and associated links
- [x] Add a way to easily collapse the first few error messages and switch to the "Errors" tab with one click

![errors-mockup-edit](https://user-images.githubusercontent.com/1713505/43736788-e8fc21ac-998c-11e8-9104-0be16e4b1687.png)

calbach commented on api/jobs.yaml:
```diff
@@ -573,13 +573,25 @@ definitions:
+      stderr:
+        type: string
+        description: Path to the standard error file for this task or shard
+      callRoot:
```
This language might be obvious for Cromwell users, but in terms of the general jobs API I have no idea what a "callRoot" is and what I should expect to find here. Is there any way we can encode this information more generally e.g. by renaming this?

rsasch replied:
"callRoot" was introduced to `jobs.yaml` in #271 (see https://github.com/DataBiosphere/job-manager/blob/master/api/jobs.yaml#L558)

calbach commented on api/jobs.yaml:
```diff
@@ -573,13 +573,25 @@ definitions:
     properties:
+      taskName:
+        type: string
+        description: The name of the task that generated the failure message
```
This should specify "if any" - This object is referenced both on the task and on the job, and dsub doesn't have tasks.

rsasch replied:
done

calbach commented on ui/src/app/core/local-storage.service.ts:
```diff
@@ -0,0 +1,5 @@
+import {Injectable} from "@angular/core";
+
+@Injectable()
+export class LocalStorageService {
```
Did you mean to include this?

rsasch commented:
@calbach The description was in the linked issue; I've moved it to the PR.

samanehsan commented on servers/cromwell/jobs/controllers/jobs_controller.py:
```diff
@@ -100,7 +100,9 @@ def get_job(id, **kwargs):
-            FailureMessage(failure=f['message']) for f in job['failures']
+            format_failure(name, m)
+            for name, metadata in job.get('calls', {}).items()
+            for m in metadata if m.get('failures') is not None
```
Wow! I've actually never seen a nested list-comprehension in python before 😮

samanehsan commented:
As discussed in person, the subworkflow detail pages are not showing the error panel & tab but there is error information in the metadata about failed tasks:

<img width="1650" alt="screen shot 2018-08-07 at 3 35 05 pm" src="https://user-images.githubusercontent.com/6414394/43798838-cac2cf92-9a59-11e8-80f9-da1647305c37.png">

bfcrampton commented on ui/src/app/job-details/failures/failures-tab/failures-tab.component.html:
```diff
@@ -0,0 +1,49 @@
+          </clr-tooltip-content>
+        </a>
+      </clr-tooltip>
+      <clr-icon *ngIf="!getResourceUrl(f.stdout)" shape="file" size="24" style="color: #ccc;"></clr-icon>
```
Per my other comment about logs on `FailureMessage` this should extract the correct log file out of the `job` or `task` (if a taskName is present).

rsasch commented:
@samanehsan I believe I've resolved the subworkflow issue; please pull the latest code and let me know.

samanehsan commented:
Thanks @rsasch! I see the error messages for the task that failed in the subworkflow now 🎉 

<img width="1562" alt="screen shot 2018-08-07 at 5 44 25 pm" src="https://user-images.githubusercontent.com/6414394/43804382-d2ef81be-9a69-11e8-91c1-681ddd02c1ec.png">



And the errors tab: 
<img width="1561" alt="screen shot 2018-08-07 at 5 45 30 pm" src="https://user-images.githubusercontent.com/6414394/43804386-d619a8d8-9a69-11e8-92f8-b5a68e3e98e2.png">

rsasch commented:
@bfcrampton Thanks for the feedback; I'm going to look into whether or not it makes sense to pass `stdout`, `stderr` and `callRoot` per failure and on making the UI more flexible when it comes to displayed columns. I will ping you for a re-review once I do.

rsasch commented:
@bfcrampton I've moved the (renamed) `failure-table` component to a `common` directory under `job-details`, added some additional inputs and now call it in both the failures panel and the failures task tab. I've also added logic in the UI which should hide the "task name" and "links" columns if there is no data in them. And I discussed the passing of the `stdout`, `stderr` and `callRoot` per failure with @ruchim; we both think that it makes sense for the JM to rely on the Cromwell backend to supply these paths, as the location scheme might change in the future and this keeps that logic in only one place. I look forward to your feedback.

rsasch commented:
@bfcrampton Thanks for the feedback; the list of errors for a task will include all shards and attempts, if applicable.
